ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/utils.kt

```diff
@@ -142,20 +145,15 @@ internal suspend fun writeBody(
     }
 }
 
-private suspend fun processOutgoingContent(
-    request: HttpRequestData,
-    body: OutgoingContent,
-    channel: ByteWriteChannel
-): Boolean {
+private suspend fun processOutgoingContent(request: HttpRequestData, body: OutgoingContent, channel: ByteWriteChannel) {
     when (body) {
-        is OutgoingContent.NoContent -> return false
         is OutgoingContent.ByteArrayContent -> channel.writeFully(body.bytes())
         is OutgoingContent.ReadChannelContent -> body.readFrom().copyAndClose(channel)
         is OutgoingContent.WriteChannelContent -> body.writeTo(channel)
-        is OutgoingContent.ProtocolUpgrade -> throw UnsupportedContentTypeException(body)
         is OutgoingContent.ContentWrapper -> processOutgoingContent(request, body.delegate(), channel)
+        is OutgoingContent.ProtocolUpgrade -> error("unreachable code")
+        is OutgoingContent.NoContent -> error("unreachable code")
```

Comment:
I wonder if you inlined this function if the compiler would be able to figure this out... maybe with the 2.0 compiler 🤔.  I'll check on it when I do the upgrade.